### ISSUE TITLE
declaring extern type containing a C fixed-size array

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#11996]
As part of resolving issue #11939, I've come upon a question about the best practice for declaring external C arrays with fixed size.

Suppose you have a data type such as this:
``` c
  typedef struct {
    double dat[2];
  } mypair;
```

What is the best practice way to declare this in Chapel with an `extern type`?

I can see two options but I like neither of them.

1. Declare it as a tuple, as suggested in 
https://chapel-lang.org/docs/technotes/extern.html#declaring-external-c-types - e.g. 
``` chapel
extern record mycomplex {
  var dat : 2*c_double;
}
```

2. Declare it as a pointer, which is what c2chapel currently does in this case, e.g.
``` chapel
extern record mycomplex {
  var dat : c_ptr(c_double);
}
```

The problem with the tuple strategy is that the indices now start at 1, which I think would be confusing when using the C interoperability (since pointer calculations / "arrays" start at 0 when using C interop). Wouldn't it be confusing if fixed-sized arrays started at 1 but other C arrays started at 0?

The problem with the pointer strategy is that various parts of the Chapel compiler / module code could benefit from knowing that `dat` is fixed size:
 * `writeln` of the type prints the pointer rather than the elements
 * The compiler-generated copy initializer for `mycomplex` tries to initialize `dat` as a pointer to the other `dat` but this doesn't make sense.
 * One might like to be able to copy a fixed-size array into another variable (but you can't do this in C, see below).
 * one would like to be able to have a compiler-generated default initializer for `mycomplex` but that's not possible if the compiler assumes `dat` is just a pointer.

Note though that C really does consider these fixed-size arrays as pointers in some ways.
``` c
#include <stdio.h>

int main() {
  int a[2];
  a[0] = 1;
  a[1] = 2;

  {
    int b[2] = a; // error - invalid initializer
    printf("%i %i\n", b[0], b[1]);
  }

  {
    int c[2];
    c = a; // error - assignment to expression with array type
    printf("%i %i\n", c[0], c[1]);
  }

  return 0;
}
```

I can see the following possible strategies to improve the situation:

 1. Introduce a fixed-size array type, e.g. c_array, and use that in this case. It would:
    * use a param size
    * be a value type (like tuples)
    * perform bounds checking when given compile time-known indices
    * only ever be 1 dimensional
    * only ever store 1 element type (unlike tuples)
    * always have the first element at index 0 (unlike tuples)
 2. Introduce a variant of tuples that allows indexing starting at 0.
 3. Adjust the compiler to use C structure assignment or memcpy for `extern record`s by default, instead of trying to assign only the fields known to the Chapel compiler.

I think (3) might be worth doing anyway, and it also has the advantage of addressing variable-sized arrays in C, but it I wonder if we could/should do better.

But, even if we attempted none of the strategies to improve the situation, we probably should at the very least update the C interoperability technote to discuss these alternatives.

**Comment by npadmana** [comment]
@mppf -- I would think that the `c2chapel` strategy could be confusing if one was building arrays of these, where the arrays would have very different memory layouts. The use case I had was the GSL definition of complex numbers, which used a fixed size array, but there are cases where you want the array of these complex numbers to be contiguous.

**Comment by bradcray** [comment]
Note the relationship between this issue and #10162.

**Comment by bradcray** [comment]
Strategies 3 and 1 are both appealing to me, 2 is less so (as is using tuples to represent such types, long-term).

At one point we discussed whether there would be a potential option 4 in which we treated such external arrays as though they were Chapel arrays with a "I'm a C array" domain map, but there are obvious challenges when the variable is a field in a struct that we don't have control over.  When wrestling with this, we pondered whether something could be done to look up the Chapel notion of the array by hashing on the base pointer value, but I don't think we ever got a solution that seemed plausible and in the few minutes I've thought about it here, it seems hard to pull off.
